subsys/net/lib/sockets/sockets_service.c

```diff
@@ -0,0 +1,286 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(net_sock_svc, CONFIG_NET_SOCKETS_LOG_LEVEL);
+
+#include <zephyr/kernel.h>
+#include <zephyr/net/socket_service.h>
+#include <zephyr/posix/sys/eventfd.h>
+
+static int init_socket_service(void);
+static bool init_done;
+
+static K_MUTEX_DEFINE(lock);
+static K_CONDVAR_DEFINE(wait_start);
+
+STRUCT_SECTION_START_EXTERN(net_socket_service_desc);
+STRUCT_SECTION_END_EXTERN(net_socket_service_desc);
+
+static struct service {
+	/* The +1 is for triggering events from register function */
+	struct zsock_pollfd events[1 + CONFIG_NET_SOCKETS_POLL_MAX];
+	int count;
+} ctx;
+
+#define get_idx(svc) (*(svc->idx))
+
+void net_socket_service_foreach(net_socket_service_cb_t cb, void *user_data)
+{
+	STRUCT_SECTION_FOREACH(net_socket_service_desc, svc) {
+		cb(svc, user_data);
+	}
+}
+
+static void cleanup_svc_events(const struct net_socket_service_desc *svc)
+{
+	for (int i = 0; i < svc->pev_len; i++) {
+		ctx.events[get_idx(svc) + i].fd = -1;
+		svc->pev[i].event.fd = -1;
+		svc->pev[i].event.events = 0;
+	}
+}
+
+int z_impl_net_socket_service_register(const struct net_socket_service_desc *svc,
+				       struct zsock_pollfd *fds, int len,
+				       void *user_data)
+{
+	int i, ret = -ENOENT;
+
+	k_mutex_lock(&lock, K_FOREVER);
+
+	if (!init_done) {
+		(void)k_condvar_wait(&wait_start, &lock, K_FOREVER);
+	}
+
+	if (STRUCT_SECTION_START(net_socket_service_desc) >= svc &&
+	    STRUCT_SECTION_END(net_socket_service_desc) < svc) {
+		goto out;
+	}
```

Comment:
This doesn't seem right.

```suggestion
	if (STRUCT_SECTION_START(net_socket_service_desc) > svc ||
	    STRUCT_SECTION_END(net_socket_service_desc) <= svc) {
		goto out;
	}
```

Reply:
Indeed, will fix.